src/crypto/NoteEncryption.cpp

```diff
@@ -0,0 +1,161 @@
+
+    // Create the ephemeral keypair
+    randombytes_buf(esk.begin(), crypto_scalarmult_SCALARBYTES);
+    epk = generate_pubkey(esk);
+}
+
+template<size_t MLEN>
+NoteEncryption<MLEN>::~NoteEncryption() {
+    sodium_memzero(epk.begin(), crypto_scalarmult_BYTES);
+    sodium_memzero(esk.begin(), crypto_scalarmult_SCALARBYTES);
+}
+
+template<size_t MLEN>
+typename NoteEncryption<MLEN>::Ciphertext NoteEncryption<MLEN>::encrypt
+                                          (const uint256 &pk_enc,
+                                           const NoteEncryption<MLEN>::Plaintext &message
+                                          )
+{
+    uint256 dhsecret;
+
+    if (crypto_scalarmult(dhsecret.begin(), esk.begin(), pk_enc.begin()) != 0) {
+        throw std::logic_error("Could not create DH secret");
+    }
+
+    // Construct the symmetric key
+    unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE];
+    KDF(K, dhsecret, epk, pk_enc, seed, nonce);
+
+    // Increment the number of encryptions we've performed
+    nonce++;
+
+    // The nonce is zero for our purposes
+    unsigned char cipher_nonce[crypto_aead_chacha20poly1305_NPUBBYTES] = {};
+
+    NoteEncryption<MLEN>::Ciphertext ciphertext;
+
+    crypto_aead_chacha20poly1305_encrypt(ciphertext.begin(), NULL,
+                                         message.begin(), MLEN,
+                                         NULL, 0, // no "additional data"
+                                         NULL, cipher_nonce, K);
+
+    return ciphertext;
+}
+
+template<size_t MLEN>
+typename NoteEncryption<MLEN>::Plaintext NoteEncryption<MLEN>::decrypt
+                                         (const uint256 &sk_enc,
+                                          const NoteEncryption<MLEN>::Ciphertext &ciphertext,
+                                          const uint256 &epk,
+                                          const uint256 &seed,
+                                          unsigned char nonce
+                                         )
+{
+    uint256 pk_enc = generate_pubkey(sk_enc);
+
+    uint256 dhsecret;
+
+    if (crypto_scalarmult(dhsecret.begin(), sk_enc.begin(), epk.begin()) != 0) {
+        throw std::logic_error("Could not create DH secret");
+    }
+
+    unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE];
+    KDF(K, dhsecret, epk, pk_enc, seed, nonce);
+
+    // The nonce is zero for our purposes
+    unsigned char cipher_nonce[crypto_aead_chacha20poly1305_NPUBBYTES] = {};
+
+    NoteEncryption<MLEN>::Plaintext plaintext;
+
+    if (crypto_aead_chacha20poly1305_decrypt(plaintext.begin(), NULL,
+                                             NULL,
+                                             ciphertext.begin(), NoteEncryption<MLEN>::CLEN,
+                                             NULL,
+                                             0,
+                                             cipher_nonce, K) != 0) {
+        throw std::runtime_error("Could not decrypt message");
```

Comment:
We ought not to crash when this fails, since if we do, validating an invalid transaction (or just one that wasn't encrypted to us) could cause the whole node to crash. Is this the best way to signal that to the caller?